pytorch_pfn_extras/profiler/_tracing.py

```diff
@@ -0,0 +1,230 @@
+import contextlib
+import json
+import os
+import threading
+import time
+from typing import Any, Dict, Generator, List, Optional, Type, Union, cast
+
+import torch
+from pytorch_pfn_extras.profiler import _util
+from pytorch_pfn_extras.writing import Writer
+
+
+class Tracer:
+    @contextlib.contextmanager
+    def add_event(self, name: str) -> Generator[None, None, None]:
+        raise NotImplementedError("Tracers must implement add_event")
+
+    def add_remote_event(self, name: str, value: Any) -> None:
+        raise NotImplementedError("Tracers must implement add_remote_event")
+
+    def clear(self) -> None:
+        raise NotImplementedError("Tracers must implement clear")
+
+    def flush(self, filename: str, writer: Writer) -> None:
+        raise NotImplementedError("Tracers must implement flush")
+
+    def enable(self, enable_flag: bool) -> None:
+        raise NotImplementedError("Tracers must implement enable")
```

Comment:
The Tracer class should have a `finalize()` or `synchronize()` interface. In particular, if the QueueWorker is not synchronized, the possibility arises that the program will terminate before all profiles are output.

Reply:
Added a finalize method, also enforced synchronization in `flush` method. Good catch!!!